src/state_awset.erl

```diff
@@ -0,0 +1,468 @@
+        state_causal_type:insert_object(Elem,
+                                        {dot_set, DeltaDotCloud1},
+                                        DeltaDataStore0),
+
+    DotCloud = state_causal_type:insert_dot_context(NewDotContext, DotCloud0),
+
+    {ok, {DataStore0, DotCloud}, {DeltaDataStore, DeltaDotCloud}}.
+
+%% @private
+remove_elem_delta(Elem, DataStore, {DeltaDataStore0, DeltaDotCloud0}) ->
+    {ok, SubDataStore} = state_causal_type:get_sub_data_store(Elem, DataStore),
+    case state_causal_type:is_bottom_data_store(SubDataStore) of
+        false ->
+            DeltaDotCloud = state_causal_type:merge_dot_clouds(
+                              DeltaDotCloud0,
+                              state_causal_type:get_dot_cloud(SubDataStore)),
+            {ok, DataStore, {DeltaDataStore0, DeltaDotCloud}};
+        true ->
+            {error, {precondition, {not_present, Elem}}}
+    end.
+
+%% @private
+remove_elems_delta([], DataStore, Delta) ->
+    {ok, DataStore, Delta};
+remove_elems_delta([Elem|Rest], DataStore, {DeltaDataStore0, DeltaDotCloud0}) ->
+    case remove_elem_delta(Elem, DataStore, {DeltaDataStore0, DeltaDotCloud0}) of
+        {ok, DataStore, Delta} -> remove_elems_delta(Rest, DataStore, Delta);
+        Error         -> Error
+    end.
+
+%% ===================================================================
+%% EUnit tests
+%% ===================================================================
+-ifdef(TEST).
+
+new_test() ->
+    ?assertEqual({?TYPE, {{{dot_map, dot_set}, orddict:new()}, ordsets:new()}},
+                 new()).
+
+query_test() ->
+    Set0 = new(),
+    Set1 = {?TYPE, {{{dot_map, dot_set}, [{<<"a">>, {dot_set, [{a, 2}]}}]},
+                    [{a, 1}, {a, 2}]}},
+    ?assertEqual(sets:new(), query(Set0)),
+    ?assertEqual(sets:from_list([<<"a">>]), query(Set1)).
+
+delta_add_test() ->
+    Actor = 1,
+    Set0 = new(),
+    {ok, {?TYPE, {delta, Delta1}}} = delta_mutate({add, <<"a">>}, Actor, Set0),
+    Set1 = merge({?TYPE, Delta1}, Set0),
+    {ok, {?TYPE, {delta, Delta2}}} = delta_mutate({add, <<"a">>}, Actor, Set1),
+    Set2 = merge({?TYPE, Delta2}, Set1),
+    {ok, {?TYPE, {delta, Delta3}}} = delta_mutate({add, <<"b">>}, Actor, Set2),
+    Set3 = merge({?TYPE, Delta3}, Set2),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set}, [{<<"a">>, {dot_set, [{1, 1}]}}]},
+                          [{1, 1}]}},
+                 {?TYPE, Delta1}),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set}, [{<<"a">>, {dot_set, [{1, 1}]}}]},
+                          [{1, 1}]}},
+                 Set1),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set},
+                           [{<<"a">>, {dot_set, [{1, 1}, {1, 2}]}}]},
+                          [{1, 1}, {1, 2}]}},
+                 {?TYPE, Delta2}),
+    ?assertEqual({?TYPE, {{{dot_map, dot_set},
+                           [{<<"a">>, {dot_set, [{1, 1}, {1, 2}]}}]},
+                          [{1, 1}, {1, 2}]}},
+                 Set2),
```

Comment:
With the correct Delta2 the Set2 will no longer show {1,1} as a support dot for "a".